homebridge: init at 1.11.0; homebridge-config-ui-x: init at 5.1.0; nixos/homebridge: init

fmoda3 opened the pull request:
## Description of changes

- Added Homebridge package (service for running homebridge)
- Added Homebridge Config UI (service for configuring homebridge)
- Added a nixos service for running both homebridge and the config ui

Decided to allow a mixture of immutable settings via nix options, and mutable changes from homebridge config ui.  On service start, the config file will forcibly overwrite fields that were set in the nix config.  Other settings will be left as is.  I took this inspiration from the Adguard Home service, which tries to merge nix options into existing adguard yml settings.  I used jq to do the json substitution.  I'm not sure if there is a better way to do it.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

fmoda3 committed:
homebridge: init at 1.11.0

fmoda3 committed:
homebridge-config-ui-x: init at 5.1.0

fmoda3 committed:
nixos/homebridge: init

cawilliamson commented:
This is an awesome start - thanks for doing all of this!

I'm trying to use your code on my server in order to setup homebridge and control a Nest thermostat but the problem I'm facing is when I try to download the Homebridge Nest plugin it fails to install.

I'd be eternally grateful if you could take a look in to this :)

fmoda3 commented:
> Just installed this. First plugin I installed worked straight out of the box: `homebridge-ring`. At first it looked like it crashed while installing, but after logging out and back in the plugin was available and functioning.

That's great to hear!  My last few pushes from a couple months ago added `bash` to the `extraPackages` per the suggestion above, as that also fixes the issue I had with installing the homebridge unifi protect plugin.

cawilliamson commented:
It's a real shame that this hasn't been merged yet.

Certainly gets a +1 from me!

eric-peiffer commented:
I'm really interested in integrating this package (nix-Darwin + home-manager setup). How can I help?

fmoda3 commented:
Just updated to latest versions again.

@cawilliamson Posted this to Discord, see if I can get some traction on it!

@eric-peiffer So, to get this in nix-darwin, we'd need to:
1. Get this PR merged.
2. Get the 2 packages building on Darwin.  A quick test on my end worked for `homebridge`, but it failed for `homebridge-config-ui-x` due to some issues with nodepty.
3. Add a launch daemon service in nix-darwin for homebridge, similar to the systemd one here.

eric-peiffer commented:
@fmoda3 I submitted a [PR](https://github.com/fmoda3/nixpkgs/pull/2) for the Darwin support (tested on aarch64).

fmoda3 commented:
@eric-peiffer Thanks for figuring it out!  I applied your changes directly to the commits in this PR, and test it on my x86_64 darwin machine.

kfiz commented:
Posted a review request to `#review-requestr:nixos.org`. Let's see if somebody looks at this. Would be good to have this merged.

pta2002 commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,65 @@
+    hash = "sha256-28jw+w/WLd7lR9ewmvA033D4Fj22qIWe8aUu23oLjLs=";
+  };
+in
+buildNpmPackage rec {
```
Unless I'm mistaken, you don't need `rec`

pta2002 commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,65 @@
+  src = fetchFromGitHub {
+    owner = "homebridge";
+    repo = "homebridge-config-ui-x";
+    rev = "v${version}";
```
Use `tag` instead (this also goes for the homebridge package):

```suggestion
    tag = "v${version}";
```

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+{
+  options.services.homebridge = with types; {
+
+    # Basic Example
+    # {
+    #   services.homebridge = {
+    #     enable = true;
+    #     # Necessary for service to be reachable
+    #     openFirewall = true;
+    #     # Most accessories need unauthenticated access
+    #     allowInsecure = true;
+    #   };
+    # }
```
Put this as the actual `example` in the bridge `mkOption`

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+    bridge = mkOption {
+      description = "Homebridge Bridge options";
+      default = { };
+      type = submodule {
```
For both this and the UI settings, I would suggest you use the `freeformType` setting here, to make your life easier when converting to JSON. It also means that this can expand without you having to manually add settings here for everything.

[Here's a recent example I did](https://github.com/NixOS/nixpkgs/blob/32ba344ba1bcf64ebfeaf742f88aa3f0c581e9fa/nixos/modules/services/torrent/cross-seed.nix#L45-L96)

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+            description = ''
+              The path to an image that should be used on the UI login page. If this is not defined the default login wallpaper will be used.
+            '';
+          };
+
+          scheduledBackupDisable = mkOption {
+            type = nullOr bool;
+            default = null;
+            description = ''
+              Set to true to disable the daily scheduled backups.
+            '';
+          };
+
+          scheduledBackupPath = mkOption {
+            type = nullOr str;
+            default = null;
+            description = ''
+              The full path to an existing directory where the automated daily backup archives
+              should be saved. The "homebridge" user MUST have read / write access to this directory.
+            '';
+          };
+
+          debug = mkOption {
+            type = nullOr bool;
+            default = null;
+            description = ''
+              Enable debug logging
+            '';
+          };
+
+          disableServerMetricsMonitoring = mkOption {
+            type = nullOr bool;
+            default = null;
+            description = ''
+              When enabled, the Homebridge UI will not collect or report cpu / memory stats
+            '';
+          };
+        };
+      };
+    };
```
The vast majority of these, with the exception of `port`, seem to have no need of being hardcoded options here, so I will once again suggest you use freeform type. Makes for a cleaner service config, and makes your life way easier in the future.

I'm not too familiar with homebridge so this is just a suggestion, but I think the `ui`, `bridge`, `allowInsecure`, `userStoragePath` and `pluginPath` could all be under a larger, freeform, `settings` option, setting defaults where it makes sense (e.g. username, port, store paths).

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+                  log: {
+                    method: "systemd",
+                    service: "homebridge"
+                  }
+                  ${optionalString (cfg.ui.host != null) '',"host": "${cfg.ui.host}"''}
+                  ${optionalString (cfg.ui.proxyHost != null) '',"proxyHost": "${cfg.ui.proxyHost}"''}
+                  ${optionalString (cfg.ui.auth != null) '',"auth": "${cfg.ui.auth}"''}
+                  ${optionalString (
+                    cfg.ui.sessionTimeout != null
+                  ) '',"sessionTimeout": ${toString cfg.ui.sessionTimeout}''}
+                  ${optionalString (cfg.ui.ssl != null) ''
+                    ,"ssl": {
+                      "key": "${cfg.ui.ssl.key}",
+                      "cert": "${cfg.ui.ssl.cert}",
+                      "pfx": "${cfg.ui.ssl.pfx}",
+                      "passphrase": "${cfg.ui.ssl.passphrase}"
+                    }''}
+                  ${optionalString (cfg.ui.tempUnits != null) '',"tempUnits": "${cfg.ui.tempUnits}"''}
+                  ${optionalString (cfg.ui.theme != null) '',"theme": "${cfg.ui.theme}"''}
+                  ${optionalString (cfg.ui.lang != null) '',"lang": "${cfg.ui.lang}"''}
+                  ${optionalString (
+                    cfg.ui.loginWallpaper != null
+                  ) '',"loginWallpaper": "${cfg.ui.loginWallpaper}"''}
+                  ${optionalString (
+                    cfg.ui.scheduledBackupDisable != null
+                  ) '',"scheduledBackupDisable": ${toString cfg.ui.scheduledBackupDisable}''}
+                  ${optionalString (
+                    cfg.ui.scheduledBackupPath != null
+                  ) '',"scheduledBackupPath": "${cfg.ui.scheduledBackupPath}"''}
+                  ${optionalString (cfg.ui.debug != null) '',"debug": ${toString cfg.ui.debug}''}
+                  ${optionalString (
+                    cfg.ui.disableServerMetricsMonitoring != null
+                  ) '',"disableServerMetricsMonitoring": ${toString cfg.ui.disableServerMetricsMonitoring}''}
+                }
+              else
+                .
+              end
+            )
+        EOF
+          )
```
If you do my `freeformType` suggestion, this is also now completely unnecessary. Just use the json writer, and then keep your later config merger to merge the two files.

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+          mv "${cfg.userStoragePath}/config.json.tmp" "${cfg.userStoragePath}/config.json"
+        fi
+
+        # Set proper permissions
+        chown homebridge "${cfg.userStoragePath}/config.json"
+        chgrp homebridge "${cfg.userStoragePath}/config.json"
+        chmod 600 "${cfg.userStoragePath}/config.json"
+        mkdir -p "${cfg.pluginPath}"
+        chown homebridge "${cfg.pluginPath}"
+        chgrp homebridge "${cfg.pluginPath}"
```
You can use `install` to do all of these in one/two operations.

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+      };
+    };
+
+    # Create a user whose home folder is the user storage path
+    users.users.homebridge = {
+      home = cfg.userStoragePath;
+      createHome = true;
+      group = "homebridge";
+      # Necessary so that this user can run journalctl
+      extraGroups = [ "systemd-journal" ];
+      isSystemUser = true;
+    };
```
Add `user`/`group` options to configure this (and use that for the `username` setting in the Homebridge config JSON, if that's what it's for).

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+
+    users.groups.homebridge = { };
+
+    # Need passwordless sudo for a few commands
+    # homebridge-config-ui-x needs for some features
+    security.sudo.extraRules = [
+      {
+        users = [ "homebridge" ];
+        commands = [
+          {
+            # Ability to restart homebridge services
+            command = "${pkgs.systemd}/bin/systemctl restart homebridge homebridge-config-ui-x";
+            options = [ "NOPASSWD" ];
+          }
+          {
+            # Ability to shutdown server
+            command = "${pkgs.systemd}/bin/shutdown -h now";
+            options = [ "NOPASSWD" ];
+          }
+          {
+            # Ability to restart server
+            command = "${pkgs.systemd}/bin/shutdown -r now";
+            options = [ "NOPASSWD" ];
+          }
+        ];
+      }
+    ];
```
This seems like it could potentially be dangerous, but unfortunately I'm not too certain of how this could be sandboxed. Although the impact here is pretty limited so it's probably (?) fine? You probably don't need to change anything.

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,756 @@
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    systemd.services.homebridge = {
```
nitpick: extra line

pta2002 reviewed (changes requested):
Sorry for the more or less rambly review here. I think in general you just need to use `freeformType` for the homebridge service and it'll make your (and user's) life way easier in the long run, since you won't need to worry about keeping the nix settings in sync with upstream.

As a bonus, it'll considerably simplify this!

Extra comments:
- Don't forget to add homebridge to the release notes (IIRC this is actually necessary for this to get merged, the others here can just be later improvements but it's always better to have it "right" from the start to avoid breaking configs.)
- Look into systemd's DynamicUser setting. Could potentially mean you don't need a `user`/`group` setting at all, if you can find a way around needing sudo for some things.
- Also look into the `StateDirectory` settings, which could mean you don't need to worry about ensuring the `/var/lib/homebridge` directories exist
- Not sure if it's the case, but if the `advertiser` setting for the bridge requires extra packages to be installed, could be nice to install them automatically.

Sorry once again. This should not be that much work to change though, hopefully. It's mostly a matter of removing code. If you want to take this to an even better level, look into adding a nixos test! Could, for now, just check that the homebridge HTTP UI is pingable, lots of examples of that in the nixos/tests directory.

kfiz commented:
@fmoda3 do you want to take a stab at the review suggestions?

fmoda3 commented:
@pta2002 Thanks for the review!  Taking a stab at these changes.  Might take me a few days, as I may not get a free moment until next week.

pta2002 commented:
No worries, I'm away for the week and won't be able to re-review until I'm home anyway, but I'm sure someone else can take care of it if needed.

fmoda3 commented:
Ok, I took a stab and making changes per the review comments!

- I made the user/group configurable in the service configuration.  I used your cross-seed configuration setup as a model for this.  I didn't switch to `DynamicUser`, because as far as I can tell, it wouldn't be able to use `sudo` for the restart operations.
- I switched to using `StateDirectory`, so that i don't need `createHome` in the user
- Removed usage of `rec` and swapped `ref` for `tag` in the packages.
- I added a note in the release notes.
- I didn't add advertiser packages, as the advertiser option was optional, and was removed in favor of the `freeFormType `submodule type.
- Replaced `mkdir/chmod/chgrp` with `install`.
- I wasn't sure where to put the example config comment.  I didn't think it belonged on the `bridge` option, nor perhaps even the `settings` option, as the example is for the root config level.  Is there a better place to put it?
- I replaced all of the service's json setting config options with `settings` and `uiSettings` fields that use the json `freeFormType` as requested.  I wasn't able to combine them into one singular `settings`, due to how homebridge's json config works.  The `platforms` field is an array of custom objects, and it seems the only requirements is that they have a `name` and `platform` field, and that they need to be unique by their `platform` field.  But, the rest of their fields are custom to the platform plugin.

For example:
```json
{
	...,
	"platforms": [
		{
            "homebridgePackagePath": "/nix/store/85bvy5zm575pan6787f0m0lwahchcsxb-homebridge-1.9.0/lib/node_modules/homebridge",
            "log": {
                "method": "systemd",
                "service": "homebridge"
            },
            "name": "Config",
            "platform": "config",
            "port": 8581,
            "restart": "sudo -n systemctl restart homebridge homebridge-config-ui-x",
            "standalone": true,
            "sudo": false,
            "tempUnits": "f"
        },
        {
            "controllers": [],
            "name": "UniFi Protect",
            "platform": "UniFi Protect"
        },        
    ],
	...
}
```

I need to ensure that the `config` platform exists and is configured correctly, while also allowing for other platforms.  It was easier to define `uiSettings` separately, and merge it into the `platforms` array afterwards.  This perhaps even makes logical sense, since homebridge and it's UI are run as separate services here.  But in any case, this did remove all the optional parameters, and the merge filter is much smaller.

I think potentially the only remaining things are:
- Figure out if there is a better place for the example config comment
- Write some nixos tests.

fmoda3 commented:
Just pushed another change that adds some nixos tests!

nfgrep commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+      '';
+    };
+
+    userStoragePath = mkOption {
```
Nit: could this just be `storagePath` ? Maybe there's a reason its prefixed with "user" that I don't know (I'm not at all familiar with the homebridge codebase so happy to be corrected), but in my head this is just "the spot homebridge stores all its stuff" not "the spot some user stores their homebridge stuff".

fmoda3 replied:
I think that's definitely what this field is intending to do inside homebridge, so I'd be happy to change it.

But, I did name it this because it's passed to the `-U` or `--user-storage-path` argument of homebridge: https://github.com/homebridge/homebridge/blob/8a71b77332f943483dc9e723af91ef7921f4925b/src/cli.ts#L49.  Seems like, it's immediately set to a field `customStoragePath`, and then later stored as `storagePath` statically on the `User` object.

nfgrep replied:
>But, I did name it this because it's passed to the -U or --user-storage-path argument of homebridge

Ah yea makes sense to keep this consistent with the CLI UX :+1:

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+          install -D -m 600 -o ${cfg.user} -g ${cfg.group} "${defaultConfigFile}" "${cfg.userStoragePath}/config.json"
+        fi
+
+        # Write the jq merge filter file to a temporary file
+        install -D -m 600 -o ${cfg.user} -g ${cfg.group} "${jqMergeFilterFile}" "${cfg.userStoragePath}/jqMergeFilter.jq"
+
+        # Write the nix override config to a temporary file
+        install -D -m 600 -o ${cfg.user} -g ${cfg.group} "${nixOverrideConfigFile}" "${cfg.userStoragePath}/nixOverrideConfig.json"
+
+        # Apply all nix override settings to config.json in a single jq operation
+        ${pkgs.jq}/bin/jq -s -f "${cfg.userStoragePath}/jqMergeFilter.jq" "${cfg.userStoragePath}/config.json" "${cfg.userStoragePath}/nixOverrideConfig.json" | ${pkgs.jq}/bin/jq . > "${cfg.userStoragePath}/config.json.tmp"
+        install -D -m 600 -o ${cfg.user} -g ${cfg.group} "${cfg.userStoragePath}/config.json.tmp" "${cfg.userStoragePath}/config.json"
+
+        # Remove temporary files
+        rm "${cfg.userStoragePath}/jqMergeFilter.jq"
+        rm "${cfg.userStoragePath}/nixOverrideConfig.json"
+        rm "${cfg.userStoragePath}/config.json.tmp"
```
I don't think you need to copy these to temporary files. You should be able to use e.g. `jqMergeFilterFile` directly

fmoda3 replied:
Consolidated these, to reference the files directly

pta2002 replied:
Great!

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+          };
+
+          # We're using systemd, so make sure logs is setup to pull from systemd
+          log = mkOption {
+            default = { };
+            description = "Log options for homebridge";
+            type = submodule {
+              freeformType = settingsFormat.type;
+              options = {
+                method = mkOption {
+                  type = str;
+                  default = "systemd";
+                  description = "Method to use for logging";
+                  readOnly = true;
+                };
+
+                service = mkOption {
+                  type = str;
+                  default = "homebridge";
+                  description = "Name of the systemd service to log to";
+                  readOnly = true;
+                };
+              };
+            };
+          };
```
I think you don't need to define inner submodules for these options, you should be able to do just e.g.

```suggestion
          log.method = mkOption {
            type = str;
            default = "systemd";
            description = "Method to use for logging";
            readOnly = true;
          };

          log.service = mkOption {
            type = str;
            default = "homebridge";
            description = "Name of the systemd service to log to";
            readOnly = true;
          };
```

fmoda3 replied:
Changed this here and below in `bridge`

pta2002 replied:
Great!

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+            readOnly = true;
+          };
+
+          bridge = mkOption {
+            description = "Homebridge Bridge options";
+            default = { };
+            type = submodule {
+              freeformType = settingsFormat.type;
+              options = {
+                name = mkOption {
+                  type = str;
+                  default = "Homebridge";
+                  description = "Name of the homebridge";
+                };
+
+                port = mkOption {
+                  type = port;
+                  default = 51826;
+                  description = "The port homebridge listens on";
+                };
+              };
+            };
+          };
```
Same thing about inner submodule, so this can just be `bridge.name` and `bridge.port`

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+            default = [ ];
+            apply = validatePlatforms;
+            type = listOf (submodule {
+              freeformType = settingsFormat.type;
```
(Really not sure): you might not need to define the freeform type here

fmoda3 replied:
I believe I need to set this.  I don't think `freeFormType` is inherited from the parent.  I did find another service module that specifies it on nested submodules: https://github.com/NixOS/nixpkgs/blob/1713b020bf44fbfd459638d9b13f397738da2db3/nixos/modules/services/security/canaille.nix

pta2002 replied:
Alright, really wasn't sure, best to leave it.

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+            description = "Homebridge Accessories";
+            default = [ ];
+            type = listOf (submodule {
+              freeformType = settingsFormat.type;
```
Same hehre for the freeform type. Not even sure if you need to define this option at all?

fmoda3 replied:
So, essentially what I am trying to enforce here, is that _if_ `accessories` is specified, and the user provides it with a list of accessories, that those accessories are required to have a `type` and a `name`.  Each accessory can then have any other attributes unique to that accessory type.  Would my approach here achieve this goal?

pta2002 replied:
Ah in that case I think this works, yes!

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+let
+  cfg = config.services.homebridge;
+
+  args = concatStringsSep " " (
+    [
+      "-U ${cfg.userStoragePath}"
+      "-P ${cfg.pluginPath}"
+      "--strict-plugin-resolution"
+    ]
+    ++ optionals cfg.allowInsecure [ "-I" ]
+  );
```
nit for readability: move this to where the `args` is used so it reads nicer

fmoda3 replied:
Moved into the services' exec statements.

pta2002 commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,549 @@
+    ];
+  };
+
+  defaultConfigFile = pkgs.writeTextFile {
+    name = "config.json";
+    text = builtins.toJSON defaultConfig;
+  };
```
I'm not sure if it's possible since you use the `defaultConfig`, but you might be able to do

```suggestion
  defaultConfigFile = settingsFormat.generate "config.json" defaultConfig;
```

Same goes for the `nixOverrideConfigFile`.

fmoda3 replied:
Made this change

pta2002 reviewed (changes requested):
Some relatively minor things. A lot nicer!

fmoda3 commented:
@pta2002 Thanks for the re-review!  Implemented most of your requests, and responded to a couple comments!

pta2002 reviewed (approved):
Looks good to me after the latest set of changes!

kfiz commented:
> Looks good to me after the latest set of changes!

So this can be merged now?

pta2002 commented:
I'm fine with it but I can't approve since I'm not a committer, you'll need
someone else!

A terça, 29/04/2025, 22:30, kfiz ***@***.***> escreveu:

> *kfiz* left a comment (NixOS/nixpkgs#262427)
> <https://github.com/NixOS/nixpkgs/pull/262427#issuecomment-2840294207>
>
> Looks good to me after the latest set of changes!
>
> @pta2002 <https://github.com/pta2002> So this can be merged now?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/262427#issuecomment-2840294207>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABYZLTAVOA5WEYMRA4G2IR3237VRDAVCNFSM6AAAAABR77M32OVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQNBQGI4TIMRQG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kfiz commented:
> I'm fine with it but I can't approve since I'm not a committer, you'll need someone else!

@ElvishJerricco  could you maybe look at this and merge if ok?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2396

SigmaSquadron commented:
> I'm fine with it but I can't approve since I'm not a committer, you'll need someone else!

Non-committers can approve PRs. We even have a labelling system that counts non-committer approvals.

JohnRTitor commented on pkgs/by-name/ho/homebridge/package.nix:
```diff
@@ -0,0 +1,43 @@
+  pname = "homebridge";
+  inherit version src;
+
+  nodejs = nodejs_22;
```
AFAIK nodejs is nodejs_22, so no need for this

fmoda3 replied:
Removed this from both packages

kfiz commented on nixos/modules/services/home-automation/homebridge.nix:
```diff
@@ -0,0 +1,512 @@
+      description = "Group to run homebridge as.";
+    };
+
+    openFirewall = mkOption {
+      type = bool;
+      default = false;
+      description = ''
+        Open ports in the firewall for the Homebridge web interface and service.
+      '';
+    };
+
+    allowInsecure = mkOption {
+      type = bool;
+      default = false;
+      description = ''
+        Allow unauthenticated requests (for easier hacking).
+        In homebridge's own installer, this is enabled by default.
+        Needs to be enabled if you want to control accessories:
+        https://github.com/homebridge/homebridge-config-ui-x/wiki/Enabling-Accessory-Control
+      '';
+    };
```
@JohnRTitor thanks for the additional review. @fmoda3 can you fix these?

fmoda3 commented on pkgs/by-name/ho/homebridge/package.nix:
```diff
@@ -0,0 +1,43 @@
+    cat package.json | ${pkgs.jq}/bin/jq '.scripts.clean = "rimraf lib/"' > package.json.tmp
+    mv package.json.tmp package.json
+    npm run build
+  '';
```
Thanks for this info!  As I was changing it to add the run hooks, I realized I probably only needed to do a preBuild to modify the package.json, rather than override the full buildPhase, so I made that change.

JohnRTitor commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,62 @@
+{
+  lib,
+  pkgs,
```
oof.

fmoda3 replied:
Removed pkgs in favor of direct references to the packages themselves

JohnRTitor commented on pkgs/by-name/ho/homebridge/package.nix:
```diff
@@ -0,0 +1,37 @@
+  # rimraf is already in the declared dependencies, so we just don't need to do it.
+  # This will replace "npm install rimraf && rimraf lib/" with "rimraf lib/".
+  preBuild = ''
+    cat package.json | ${pkgs.jq}/bin/jq '.scripts.clean = "rimraf lib/"' > package.json.tmp
```
```suggestion
    cat package.json | ${jq}/bin/jq '.scripts.clean = "rimraf lib/"' > package.json.tmp
```

JohnRTitor replied:
or just adding jq to nativeBuildInputs will be sufficient I believe, then you can run it with just `jq`.

JohnRTitor commented on pkgs/by-name/ho/homebridge/package.nix:
```diff
@@ -0,0 +1,37 @@
+{
+  lib,
+  pkgs,
+  buildNpmPackage,
+  fetchFromGitHub,
```
```suggestion
  buildNpmPackage,
  fetchFromGitHub,
  jq,
```

JohnRTitor commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,62 @@
+  # Uses the prebuild version of @homebridge/node-pty-prebuilt-multiarch instead
+  makeCacheWritable = pkgs.stdenv.hostPlatform.isDarwin;
+
+  nativeBuildInputs = with pkgs; [
+    python3
+    (lib.optional stdenv.hostPlatform.isDarwin cacert)
+  ];
```
```suggestion
  nativeBuildInputs = [
    python3
  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [ cacert ];
```

JohnRTitor commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,62 @@
+{
+  lib,
+  pkgs,
+  buildNpmPackage,
+  fetchFromGitHub,
+  fetchNpmDeps,
+  npmHooks,
```
```suggestion
  stdenv,
  buildNpmPackage,
  fetchFromGitHub,
  fetchNpmDeps,
  npmHooks,
  python3
```

JohnRTitor reviewed (commented):
oof, using pkgs is like this except in modules, is not recommended.

JohnRTitor commented:
> @JohnRTitor As far as I can see all change requests have been addressed. What do you think? Can we merge?

Sorry, with the amount of review requests and notifications things can get a bit lost! Feel free to ping me again if  I don't respond for a while.

JohnRTitor commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,62 @@
+  };
+
+  # Deps hash for the root package
+  npmDepsHash = "sha256-aqG9GQ2gdrCz+lxTOVYFo96jusb710KkguN/5R0Ye6c=";
+
+  # Need to also run npm ci in the ui subdirectory
+  preBuild =
+    let
+      # Deps src and hash for ui subdirectory
+      npmDeps_ui = fetchNpmDeps {
+        name = "npm-deps-ui";
+        src = "${finalAttrs.src}/ui";
+        hash = "sha256-WHGg2sINF1fH3tnuZsf+yjqoyWSPU5osOhuZb5VU+8I=";
+      };
+    in
+    ''
+      # Tricky way to run npmConfigHook multiple times
+      (
+        source ${npmHooks.npmConfigHook}/nix-support/setup-hook
+        npmRoot=ui npmDeps=${npmDeps_ui} makeCacheWritable= npmConfigHook
+      )
```
```suggestion
  npmDepsHash = "sha256-aqG9GQ2gdrCz+lxTOVYFo96jusb710KkguN/5R0Ye6c=";

  # Deps src and hash for ui subdirectory
  npmDeps_ui = fetchNpmDeps {
    name = "npm-deps-ui";
    src = "${finalAttrs.src}/ui";
    hash = "sha256-WHGg2sINF1fH3tnuZsf+yjqoyWSPU5osOhuZb5VU+8I=";
  };

  # Need to also run npm ci in the ui subdirectory
  preBuild =  
    ''
      # Tricky way to run npmConfigHook multiple times
      (
        source ${npmHooks.npmConfigHook}/nix-support/setup-hook
        npmRoot=ui npmDeps=${finalAttrs.npmDeps_ui} makeCacheWritable= npmConfigHook
      )
```

something like this.

fmoda3 commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,62 @@
+
+  # On darwin, the build failed because openpty() is not declared
+  # Uses the prebuild version of @homebridge/node-pty-prebuilt-multiarch instead
+  makeCacheWritable = pkgs.stdenv.hostPlatform.isDarwin;
```
If I don't set this on darwin, the build runs into build issues with `node-pty-prebuilt-multiarch`

```
error: builder for '/nix/store/3hy3hi7ak4hl3ajkylm58psmp4wv0292-homebridge-config-ui-x-4.75.0.drv' failed with exit code 1;
       last 25 log lines:
       > npm error In file included from /nix/store/q53z5fvh5gl581fw123nn8b5ghbllvz1-nodejs-22.14.0-source/deps/v8/include/v8-object.h:10:
       > npm error /nix/store/q53z5fvh5gl581fw123nn8b5ghbllvz1-nodejs-22.14.0-source/deps/v8/include/v8-persistent-handle.h:512:26: warning: cast from 'typename WeakCallbackInfo<ObjectWrap>::Callback' (aka 'void (*)(const WeakCallbackInfo<ObjectWrap> &)') to 'Callback' (aka 'void (*)(const WeakCallbackInfo<void> &)') converts to incompatible function type [-Wcast-function-type-mismatch]
       > npm error   512 |                          reinterpret_cast<Callback>(callback), type);
       > npm error       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > npm error ../../../nan/nan_object_wrap.h:64:50: note: in instantiation of function template specialization 'v8::PersistentBase<v8::Object>::SetWeak<Nan::ObjectWrap>' requested here
       > npm error    64 |     persistent().v8::PersistentBase<v8::Object>::SetWeak(
       > npm error       |                                                  ^
       > npm error 5 warnings and 1 error generated.
       > npm error make: *** [pty.target.mk:124: Release/obj.target/pty/src/unix/pty.o] Error 1
       > npm error gyp ERR! build error
       > npm error gyp ERR! stack Error: `make` failed with exit code: 2
       > npm error gyp ERR! stack at ChildProcess.<anonymous> (/nix/store/g5bv4gi6p7ryhs2hbaryxs6ivsxa6lqc-nodejs-22.14.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:216:23)
       > npm error gyp ERR! System Darwin 24.5.0
       > npm error gyp ERR! command "/nix/store/g5bv4gi6p7ryhs2hbaryxs6ivsxa6lqc-nodejs-22.14.0/bin/node" "/nix/store/g5bv4gi6p7ryhs2hbaryxs6ivsxa6lqc-nodejs-22.14.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
       > npm error gyp ERR! cwd /private/tmp/nix-build-homebridge-config-ui-x-4.75.0.drv-0/source/node_modules/@homebridge/node-pty-prebuilt-multiarch
       > npm error gyp ERR! node -v v22.14.0
       > npm error gyp ERR! node-gyp -v v11.0.0
       > npm error gyp ERR! not ok
       > npm notice
       > npm notice New major version of npm available! 10.9.2 -> 11.4.1
       > npm notice Changelog: https://github.com/npm/cli/releases/tag/v11.4.1
       > npm notice To update run: npm install -g npm@11.4.1
       > npm notice
       > npm error Log files were not written due to an error writing to the directory: /nix/store/vcggs0919rcmslxk8xskd7r2v69qv4xj-homebridge-config-ui-x-4.75.0-npm-deps/_logs
       > npm error You can rerun the command with `--loglevel=verbose` to see the logs in your terminal
```
Note also the `makeCacheWriteable=` part of the `preBuild` as the `ui` subdirectory also fails on darwin without it.

I saw this pattern in other packages, like here: https://github.com/NixOS/nixpkgs/blob/12943bb18fb1bb7fe6de2dd9fad4557196cff200/pkgs/by-name/in/inshellisense/package.nix

JohnRTitor replied:
I do not know of any other solution as I don't use Darwin. Let's wait a bit for emily to reply.

kfiz replied:
@fmoda3 looking at the link you provided... can't you just provide `openpty` as dependency where it's needed in the build?

kfiz replied:
@JohnRTitor I think we have to live with `makeCacheWriteable`. What do you think?

JohnRTitor replied:
Sure.

fmoda3 commented:
Just pushed an update that switches from the deprecated Standalone mode to their Service mode.  Recent versions had a warning saying that an upcoming release would completely remove Standalone mode.  I had originally used Standalone mode because it was easier to indicate where homebridge and such was.

The main difference from before, is that instead of 2 services, there is now just one that runs the `hb-service` command.  When `hb-service` runs, it creates child processes for `homebridge` and `homebridge-config-ui-x`.  It is able to find `homebridge-config-ui-x` just fine because `hb-service` exists in that package.  Unfortunately though, there is no way to tell `hb-service` where `homebridge` lives directly, and it only has a few ways to auto-discover it.  The main way is to ensure that the `homebridge` and `homebridge-config-ui-x` node modules are siblings in a directory structure, so I used some symlinking to achieve this effect and get `hb-service` to discover it.

The good news, is now this no longer uses a deprecated mode of running.

JohnRTitor commented on pkgs/by-name/ho/homebridge/package.nix:
```diff
@@ -0,0 +1,37 @@
+  '';
+
+  meta = {
+    description = "Homebridge";
```
```suggestion
    description = "Lightweight emulator of iOS HomeKit API";
```

JohnRTitor commented on pkgs/by-name/ho/homebridge-config-ui-x/package.nix:
```diff
@@ -0,0 +1,61 @@
+  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [ cacert ];
+
+  meta = {
+    description = "Homebridge Config UI X";
```
```suggestion
    description = "Configure Homebridge, monitor and backup from a browser";
```

JohnRTitor commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 262427`

Logs: https://github.com/JohnRTitor/nixpkgs-review-gha/actions/runs/16108739156


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:white_check_mark: 1 test built:</summary>
  <ul>
    <li>nixosTests.homebridge</li>
  </ul>
</details>

JohnRTitor commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 262427`

Logs: https://github.com/JohnRTitor/nixpkgs-review-gha/actions/runs/16108732606


---
### `x86_64-linux` (sandbox = true)
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>homebridge</li>
    <li>homebridge-config-ui-x</li>
  </ul>
</details>


---
### `aarch64-linux` (sandbox = true)
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 2 packages built:</summary>
  <ul>
    <li>homebridge</li>
    <li>homebridge-config-ui-x</li>
  </ul>
</details>


---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>homebridge-config-ui-x</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>homebridge</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>homebridge-config-ui-x</summary>
<pre>npm error In file included from /nix/store/6wqyvqp3dnch8jj3pvcsx8smssll67gs-nodejs-22.16.0-source/deps/v8/include/v8-array-buffer.h:13:
npm error In file included from /nix/store/6wqyvqp3dnch8jj3pvcsx8smssll67gs-nodejs-22.16.0-source/deps/v8/include/v8-object.h:10:
npm error /nix/store/6wqyvqp3dnch8jj3pvcsx8smssll67gs-nodejs-22.16.0-source/deps/v8/include/v8-persistent-handle.h:512:26: warning: cast from 'typename WeakCallbackInfo<ObjectWrap>::Callback' (aka 'void (*)(const WeakCallbackInfo<ObjectWrap> &)') to 'Callback' (aka 'void (*)(const WeakCallbackInfo<void> &)') converts to incompatible function type [-Wcast-function-type-mismatch]
npm error   512 |                          reinterpret_cast<Callback>(callback), type);
npm error       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
npm error ../../../nan/nan_object_wrap.h:64:50: note: in instantiation of function template specialization 'v8::PersistentBase<v8::Object>::SetWeak<Nan::ObjectWrap>' requested here
npm error    64 |     persistent().v8::PersistentBase<v8::Object>::SetWeak(
npm error       |                                                  ^
npm error 5 warnings and 1 error generated.
npm error make: *** [pty.target.mk:124: Release/obj.target/pty/src/unix/pty.o] Error 1
npm error gyp ERR! build error 
npm error gyp ERR! stack Error: `make` failed with exit code: 2
npm error gyp ERR! stack at ChildProcess.<anonymous> (/nix/store/swzij8xmzv1c81rvhcf2fzlnarjpjcnb-nodejs-22.16.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:216:23)
npm error gyp ERR! System Darwin 22.6.0
npm error gyp ERR! command "/nix/store/swzij8xmzv1c81rvhcf2fzlnarjpjcnb-nodejs-22.16.0/bin/node" "/nix/store/swzij8xmzv1c81rvhcf2fzlnarjpjcnb-nodejs-22.16.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm error gyp ERR! cwd /nix/var/nix-build-homebridge-config-ui-x-4.76.0.drv-0/source/node_modules/@homebridge/node-pty-prebuilt-multiarch
npm error gyp ERR! node -v v22.16.0
npm error gyp ERR! node-gyp -v v11.0.0
npm error gyp ERR! not ok
npm error A complete log of this run can be found in: /nix/var/nix-build-homebridge-config-ui-x-4.76.0.drv-0/cache/_logs/2025-07-07T06_03_17_326Z-debug-0.log</pre>
</details>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>homebridge-config-ui-x</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>homebridge</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>homebridge-config-ui-x</summary>
<pre>npm error In file included from /nix/store/q916q90nma27qyqidsl949xf43r2jgkg-nodejs-22.16.0-source/deps/v8/include/v8-array-buffer.h:13:
npm error In file included from /nix/store/q916q90nma27qyqidsl949xf43r2jgkg-nodejs-22.16.0-source/deps/v8/include/v8-object.h:10:
npm error /nix/store/q916q90nma27qyqidsl949xf43r2jgkg-nodejs-22.16.0-source/deps/v8/include/v8-persistent-handle.h:512:26: warning: cast from 'typename WeakCallbackInfo<ObjectWrap>::Callback' (aka 'void (*)(const WeakCallbackInfo<ObjectWrap> &)') to 'Callback' (aka 'void (*)(const WeakCallbackInfo<void> &)') converts to incompatible function type [-Wcast-function-type-mismatch]
npm error   512 |                          reinterpret_cast<Callback>(callback), type);
npm error       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
npm error ../../../nan/nan_object_wrap.h:64:50: note: in instantiation of function template specialization 'v8::PersistentBase<v8::Object>::SetWeak<Nan::ObjectWrap>' requested here
npm error    64 |     persistent().v8::PersistentBase<v8::Object>::SetWeak(
npm error       |                                                  ^
npm error 5 warnings and 1 error generated.
npm error make: *** [pty.target.mk:124: Release/obj.target/pty/src/unix/pty.o] Error 1
npm error gyp ERR! build error 
npm error gyp ERR! stack Error: `make` failed with exit code: 2
npm error gyp ERR! stack at ChildProcess.<anonymous> (/nix/store/abxz7wpw5kwaijzzmf8rl4ga981vk8pc-nodejs-22.16.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:216:23)
npm error gyp ERR! System Darwin 23.6.0
npm error gyp ERR! command "/nix/store/abxz7wpw5kwaijzzmf8rl4ga981vk8pc-nodejs-22.16.0/bin/node" "/nix/store/abxz7wpw5kwaijzzmf8rl4ga981vk8pc-nodejs-22.16.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
npm error gyp ERR! cwd /nix/var/nix-build-homebridge-config-ui-x-4.76.0.drv-0/source/node_modules/@homebridge/node-pty-prebuilt-multiarch
npm error gyp ERR! node -v v22.16.0
npm error gyp ERR! node-gyp -v v11.0.0
npm error gyp ERR! not ok
npm error A complete log of this run can be found in: /nix/var/nix-build-homebridge-config-ui-x-4.76.0.drv-0/cache/_logs/2025-07-07T05_56_56_456Z-debug-0.log</pre>
</details>
</details>

fmoda3 commented:
@JohnRTitor Made the changes you requested (description, removing with lib, and using default nodejs).  I see in the nixpkgs-review, that the config ui package is failing on darwin, which is odd though since it builds on my aarch64-darwin machine.

kfiz commented:
> @JohnRTitor Made the changes you requested (description, removing with lib, and using default nodejs). I see in the nixpkgs-review, that the config ui package is failing on darwin, which is odd though since it builds on my aarch64-darwin machine.

Didn't it used to work on darwin at some point. Could you do a `git blame`?

JohnRTitor commented:
```
1 johnrtitor@darwin01 ~ % nix build --print-out-paths "github:nixos/nixpkgs/refs/pull/262427/merge#homebridge-config-ui-x"
error: builder for '/nix/store/6h7dwfkxvqb22l951na7mzkhmwixw19p-homebridge-config-ui-x-4.78.1.drv' failed with exit code 1;
       last 25 log lines:
       > npm error       |                  ^
       > npm error In file included from ../src/unix/pty.cc:20:
       > npm error In file included from ../../../nan/nan.h:62:
       > npm error In file included from /nix/store/b9dlk23pg0m4i5q15izcqafbmi8gbmbx-nodejs-22.17.0-source/src/node.h:74:
       > npm error In file included from /nix/store/b9dlk23pg0m4i5q15izcqafbmi8gbmbx-nodejs-22.17.0-source/deps/v8/include/v8.h:24:
       > npm error In file included from /nix/store/b9dlk23pg0m4i5q15izcqafbmi8gbmbx-nodejs-22.17.0-source/deps/v8/include/v8-array-buffer.h:13:
       > npm error In file included from /nix/store/b9dlk23pg0m4i5q15izcqafbmi8gbmbx-nodejs-22.17.0-source/deps/v8/include/v8-object.h:10:
       > npm error /nix/store/b9dlk23pg0m4i5q15izcqafbmi8gbmbx-nodejs-22.17.0-source/deps/v8/include/v8-persistent-handle.h:512:26: warning: cast from 'typename WeakCallbackInfo<ObjectWrap>::Callback' (aka 'void (*)(const WeakCallbackInfo<ObjectWrap> &)') to 'Callback' (aka 'void (*)(const WeakCallbackInfo<void> &)') converts to incompatible function type [-Wcast-function-type-mismatch]
       > npm error   512 |                          reinterpret_cast<Callback>(callback), type);
       > npm error       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > npm error ../../../nan/nan_object_wrap.h:64:50: note: in instantiation of function template specialization 'v8::PersistentBase<v8::Object>::SetWeak<Nan::ObjectWrap>' requested here
       > npm error    64 |     persistent().v8::PersistentBase<v8::Object>::SetWeak(
       > npm error       |                                                  ^
       > npm error 5 warnings and 1 error generated.
       > npm error make: *** [pty.target.mk:124: Release/obj.target/pty/src/unix/pty.o] Error 1
       > npm error gyp ERR! build error
       > npm error gyp ERR! stack Error: `make` failed with exit code: 2
       > npm error gyp ERR! stack at ChildProcess.<anonymous> (/nix/store/gaq6v57mm40g1974hyg1vvxgzxqh16af-nodejs-22.17.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:216:23)
       > npm error gyp ERR! System Darwin 24.5.0
       > npm error gyp ERR! command "/nix/store/gaq6v57mm40g1974hyg1vvxgzxqh16af-nodejs-22.17.0/bin/node" "/nix/store/gaq6v57mm40g1974hyg1vvxgzxqh16af-nodejs-22.17.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
       > npm error gyp ERR! cwd /private/tmp/nix-build-homebridge-config-ui-x-4.78.1.drv-0/source/node_modules/@homebridge/node-pty-prebuilt-multiarch
       > npm error gyp ERR! node -v v22.17.0
       > npm error gyp ERR! node-gyp -v v11.0.0
       > npm error gyp ERR! not ok
       > npm error A complete log of this run can be found in: /private/tmp/nix-build-homebridge-config-ui-x-4.78.1.drv-0/cache/_logs/2025-07-20T16_31_21_344Z-debug-0.log
       For full logs, run:
         nix log /nix/store/6h7dwfkxvqb22l951na7mzkhmwixw19p-homebridge-config-ui-x-4.78.1.drv
nix build --print-out-paths   5.96s user 7.73s system 21% cpu 1:04.52 total
```

Tried on nix-community darwin builder.

If darwin support can't be fixed, we can mark it broken on darwin, then merge this.

fmoda3 commented:
Bit strange, cause it works for me on my M4 Mac Studio, both with and without sandbox:

```
❯ nix build --print-out-paths "github:nixos/nixpkgs/refs/pull/262427/merge#homebridge-config-ui-x"
/nix/store/j7i1p4hzbmk1v1p231k9nfv3pdc3dyaw-homebridge-config-ui-x-4.78.1
```

```
❯ sudo nix build --sandbox --print-out-paths "github:nixos/nixpkgs/refs/pull/262427/merge#homebridge-config-ui-x"
warning: $HOME ('/Users/fmoda3') is not owned by you, falling back to the one defined in the 'passwd' file ('/var/root')
/nix/store/j7i1p4hzbmk1v1p231k9nfv3pdc3dyaw-homebridge-config-ui-x-4.78.1
```

Might be a difference between a physical machine and a VM?

I believe what is meant to happen, is that the `node-pty-multiarch` library is supposed to detect and download a prebuilt binary to use for macos (hence making config cache writeable), and maybe it falls back to attempt a compile on VMs?

The same workaround is used here, which presumably works: https://github.com/NixOS/nixpkgs/blob/0847fffe35483af007f1c359e0585d05bb7adb55/pkgs/by-name/in/inshellisense/package.nix#L22

I'm personally not opposed to marking darwin as broken, as I don't deploy homebridge to a darwin machine, but others above seemed interested (someone could try to fix it in a follow up PR though)

JohnRTitor commented:
Nix community Darwin builder is not a VM it's a physical Mac Mini (https://nix-community.org/infrastructure/#community-builders) 

So I am not sure what's going on here. 

CC @emilazy @toonn

emilazy commented:
Non-FOD builds must not download from the internet under any circumstances.

> Bit strange, cause it works for me on my M4 Mac Studio, both with and without sandbox:
> 
> ```
> ❯ nix build --print-out-paths "github:nixos/nixpkgs/refs/pull/262427/merge#homebridge-config-ui-x"
> /nix/store/j7i1p4hzbmk1v1p231k9nfv3pdc3dyaw-homebridge-config-ui-x-4.78.1
> ```
> 
>  ```
> ❯ sudo nix build --sandbox --print-out-paths "github:nixos/nixpkgs/refs/pull/262427/merge#homebridge-config-ui-x"
> warning: $HOME ('/Users/fmoda3') is not owned by you, falling back to the one defined in the 'passwd' file ('/var/root')
> /nix/store/j7i1p4hzbmk1v1p231k9nfv3pdc3dyaw-homebridge-config-ui-x-4.78.1
> ```

This won’t work without `--rebuild`. Since you already built it once, Nix won’t try to build it again. So this isn’t testing the sandboxed build.

I don’t have any more context/knowledge than the above.

JohnRTitor reviewed (approved):
```
[johnrtitor@build05:~]$ nix build --print-out-paths github:nixos/nixpkgs/refs/pull/262427/merge#nixosTests.homebridge --print-out-paths
/nix/store/hxi25hlxs706a1hkdkpj131pkk6zi2d6-vm-test-run-homebridge
```

builds fine in aarch64-linux.

michaelbrusegard commented:
Hey, I am unable to get this working when I try to use it on the latest `unstable`. I get that the homebridge package is unavailable. It works if I use it directly from your branch though
